Added functionality for loading yamls and finding files

elenya-grant opened the pull request:
<!--
IMPORTANT NOTES

1. Use GH flavored markdown when writing your description:
   https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

2. If all boxes in the PR Checklist cannot be checked, this PR should be marked as a draft.

3. DO NOT DELETE ANYTHING FROM THIS TEMPLATE. If a section does not apply to you, simply write
   "N/A" in the description.

4. Code snippets to highlight new, modified, or problematic functionality are highly encouraged,
   though not required. Be sure to use proper code highlighting as demonstrated below.

   ```python
    def a_func():
        return 1

    a = 1
    b = a_func()
    print(a + b)
    ```
-->

<!--The title should clearly define your contribution succinctly.-->
# Added functionality for loading yamls and finding files

<!-- Describe your feature here. Please include any code snippets or examples in this section. -->
Added new utilities to improve flexibility, most notably these include:
1. `get_path()` and `find_file()` methods which are used to find folders or files that may have file paths defined as:
   + absolute paths
   + relative to cwd
   + relative to H2Integrate
   + relative to some input `root_folder` (only for `find_file()`)
2. Internal `load_yaml()` method which includes the `!include` functionality but is more flexible to the different possible ways of defining a filepath (instead of using the hopp load_yaml function)
3. Allowing for `config_file` input to H2IntegrateModel to be handled as either a dictionary or a filepath. This required some minor modifications to other functions (`collect_custom_models()` and `create_custom_models()`) which relied on filepaths that would only be available if `config_file` is input as a filepath.
4. Allowing for more file-placement flexibility by integrating the usage of `find_file()` and `get_path()` in `H2IntegrateModel.load_configs()`

So - main things are more flexibility in defining file paths and using a load_yaml() function that is not reliant on hopp.

## Type of Contribution
<!-- Check all that apply to help reviewers understand your contribution -->
- [x] Feature Enhancement
  - [ ] New Technology Model
- [ ] Bug Fix
- [ ] Documentation Update
- [ ] CI Changes
- [ ] Other (please describe):

## General PR Checklist

<!--Tick these boxes if they are complete, or format them as "[x]" for the markdown to render. -->
- [x] `CHANGELOG.md` has been updated to describe the changes made in this PR
- [x] Documentation
  - [x] Docstrings are up-to-date
  - [x] Related `docs/` files are up-to-date, or added when necessary
  - [x] Documentation has been rebuilt successfully
  - [-] Examples have been updated (if applicable)
- [x] Tests pass (If not, and this is expected, please elaborate in the tests section)
- [x] Added tests for new functionality or bug fixes
- [x] PR description thoroughly describes the new feature, bug fix, etc.

## New Technology Checklist
<!-- Complete this section only if you checked "New Technology Model" above -->
- [ ] **Performance Model**: Technology performance model has been implemented and follows H2Integrate patterns (if applicable)
- [ ] **Cost Model**: Technology cost model has been implemented (if applicable)
- [ ] **Tests**: Unit tests have been added for the new technology
  - [ ] Performance model tests (if applicable)
  - [ ] Cost model tests (if applicable)
  - [ ] Integration tests with H2Integrate system
- [ ] **Example**: A working example demonstrating the new technology has been created
  - [ ] Example has been tested and runs successfully in `test_all_examples.py`
  - [ ] Example is documented with clear explanations in `examples/README.md`
    - [ ] Input file comments
    - [ ] Run file comments
- [ ] **Documentation**:
  - [ ] Technology documentation page added to `docs/technology_models/`
  - [ ] Technology added to the main technology models list in `docs/technology_models/technology_overview.md`
- [ ] **Integration**: Technology has been properly integrated into H2Integrate
  - [ ] Added to `supported_models.py`
  - [ ] If a new commodity_type is added, update `create_financial_model` in `h2integrate_model.py`
  - [ ] Follows established naming conventions outlined in `docs/developer_guide/coding_guidelines.md`

## Related issues

<!--If one exists, link to a related GitHub Issue.-->
#197, #91, and kind of #6

## Impacted areas of the software

<!--
Replace the below example with any added or modified files, and briefly describe what has been changed or added, and why.
-->
- `docs/misc_resources/defining_filepaths.md`: doc page briefly explaining filepath naming
- `h2integrate/core/h2integrate_model.py`
  - `load_config()`: updated to handle if config_file is a dictionary (allowing users to input a dictionary instead of a filename), uses load_yaml() method to load config_file if its not a dictionary.
  - `create_custom_models()`: updated to use `find_file()` method to locate the custom model, specifically to handle cases where a user input a dictionary rather than config file and the `config_parent_path` input variable may not exist.
  - `collect_custom_models`: updated `config_parent_path` input to `create_custom_models()`, now uses `self.tech_path_parent` or `self.plant_path_parent` rather than `self.tech_config_path.parent` and `self.plant_config_path.parent` since these attributes may not be Path objects if a user input the config_file as a dictionary.
- `h2integrate/core/utilities.py`
  - `write_yaml()`: method moved from validation.py and updated to have more optional inputs (such as `convert_np` and `check_formatting`) 
  - `load_yaml()` and `Loader`: yaml load functions that allow for !include functionality that exists in the hopp loader, but can use the !include functionality with absolute file paths in addition to relative file paths.
  - `remove_numpy()` moved from validation.py file because only used in `write_yaml()` method
  - `get_path()`: function modified from [the function in Aviary](https://github.com/OpenMDAO/Aviary/blob/f575bc743ef1a7736031ad6687c743596424d72a/aviary/utils/functions.py#L353), used to locate a folder or file from an input path that could be absolute, relative to cwd, or relative to h2integrate package.
  - `find_file()`: new method to search for a file from an absolute path, with respect to an input root directory, current working directory, and the h2integrate package directory.
- `h2integrate/core/inputs/validation.py`: 
  - `write_yaml()` and `remove_numpy()` were moved to utilities.py
  - updated load_yaml import to import from h2integrate.core.utilities rather than hopp.
- `h2integrate/core/test/test_utilities.py`
  - `test_get_path`: new tests for `get_path()` method
  - `test_find_file`: new tests for `find_file()` method

## Additional supporting information

<!--Add any other context about the problem here.-->
More information on `find_file()`:

**This method will throw an error if it finds multiple or zero matching files in any of the places it searches.**

1. Look for file relative to a root folder (if input)
2. Look for file relative to the cwd
3. Look for file relative to the H2Integrate root (ROOT_DIR.parent)
4. Check if its a full filepath
5. Throw error if no files found in Step 2 and Step 3
6. Throw error if no files found in Step 1 (and root folder was input)
7. Throw error if multiple files were found in any of the steps

## Test results, if applicable

<!--
Add the results from unit tests and regression tests here along with justification for any
failing test cases.
-->


<!--
__ For NREL use __
Release checklist:
- [ ] Update the version in h2integrate/__init__.py
- [ ] Verify docs builds correctly
- [ ] Create a tag on the main branch in the NREL/H2Integrate repository and push
- [ ] Ensure the Test PyPI build is successful
- [ ] Create a release on the main branch
-->

johnjasa reviewed (changes requested):
I love this PR! I think it's a great value-add to the code and makes the tool more predictable for users by giving more options for supplying files to H2I. I think this is a big enough PR in terms of user experience that I want to get folks' views on it in the code standup.

I pushed up some docstring/comment additions directly. The only changes I'm requesting would be to add to the docs about the file loading behavior and options. This is a pretty impactful set of feature additions for users, so mentioning it both on its own doc page, but also elsewhere early on the docs is what I would suggest. Let me know if that's unclear or if you want to work on it together!

johnjasa commented:
> I think this is amazing functionality!!! The one thing I'd like to discuss is the find_file function:
> 
> 1. Is the order that the function looks for files what we want it to be? i.e. start with the root_directory, then look in current directory, then in H2I, then absolute (I think)
> 2. Because this function is used within H2I without the option to specify a root directory, I'm concerned that this will muddy the waters about where we are loading things in from. We tend (maybe just me) to name input files the same thing, so if it ends up looking for a driver_config.yaml file in H2I, it's not clear to me what it will find.

Great comments! I'll ask you directly: what _should_ the order be for `find_file`, in your eyes?

elenya-grant commented:
> I think this is amazing functionality!!! The one thing I'd like to discuss is the find_file function:
> 
> 1. Is the order that the function looks for files what we want it to be? i.e. start with the root_directory, then look in current directory, then in H2I, then absolute (I think)
> 2. Because this function is used within H2I without the option to specify a root directory, I'm concerned that this will muddy the waters about where we are loading things in from. We tend (maybe just me) to name input files the same thing, so if it ends up looking for a driver_config.yaml file in H2I, it's not clear to me what it will find.

Thanks for your first point - I do think this would be important to discuss with everyone else.

Regarding your second point - this is partially why I added attributes to H2IntegrateModel for these file paths (`driver_config_path`, `tech_config_path`, and `plant_config_path`) - so these could be checked to ensure the right file is being loaded. Also - if `find_file()` looks for a driver_config.yaml file relative to H2I, it will find all the driver_config.yaml files in examples/ and elsewhere and throw an error. The `find_file()` function will only return a filepath if it found a single file that matches, but throw an error if it finds multiple matches. This somewhat enforces some specific file naming practices such as:
1. Changing your cwd to the specific example or case you're running (like `cd examples/01_onshore_steel_mn`) - this is what H2I has always somewhat required to run the examples OR
2. Specifying full file paths instead of relative ones (I think this is the most error-prone way of finding files) OR
3. Naming files with unique names.

I'd say that beforehand, there was some limitations on the file handling or some code that had some "looking" logic that wasn't necessarily clear. The tech_config, plant_config, and driver_config, had to be in the same folder as the top-level config. For example, if a relative filepath is provided (for example) `"tech_inputs/hopp_config.yaml"`, and that's converted to a Path object, the absolute filepath of that Path object is with respect to the cwd. The load_yaml() function in hopp only allows the !include functionality to be used for files with respect to the file they're being included from, and would break if full file paths were provided.

elenya-grant commented:
> I love this PR! I think it's a great value-add to the code and makes the tool more predictable for users by giving more options for supplying files to H2I. I think this is a big enough PR in terms of user experience that I want to get folks' views on it in the code standup.
> 
> I pushed up some docstring/comment additions directly. The only changes I'm requesting would be to add to the docs about the file loading behavior and options. This is a pretty impactful set of feature additions for users, so mentioning it both on its own doc page, but also elsewhere early on the docs is what I would suggest. Let me know if that's unclear or if you want to work on it together!

Thanks John!! I added a note in `docs/user_guide/how_to_set_up_an_analysis.ipynb` under the 'top-level config' section that explains the file naming. I also added a simple doc page on this here: `docs/misc_resources/defining_filepaths.md`

genevievestarke commented:
Perfect, I love the description of a root folder there! I made some minor edits to the docs page, but I think it looks great!

johnjasa commented on docs/misc_resources/defining_filepaths.md:
````diff
@@ -0,0 +1,54 @@
+    driver_config: "/Users/myname/H2Integrate/examples/01_onshore_steel_mn/driver_config.yaml"
+    ```
+
+Note that if more than one file is found with the same name, then the code will give an error. A filepath will only be returned if there is one, and only one, file that matches.
````
If I'm reading this note correctly, if H2I successfully finds a file using e.g. methods 1 and 2 above, then an error is raised? I looked at the actual code and don't think I saw that logic, so I think this doc could be more clearly worded. Is this only if there's a filename overlap within _one_ of the given search methods?

johnjasa commented on h2integrate/core/h2integrate_model.py:
```diff
@@ -65,19 +63,134 @@ def __init__(self, config_file):
+
+        if isinstance(config_file, dict):
+            # Use dictionary directly without filepath context
+            config = config_file
```
Non-blocking nitpick for clarity: maybe it makes sense to rename `config_file` to something like `config_input`, which can then become the `config` dict directly. Because the name `config_file` has `file` in it, we could be more clear that it might be a dict by changing its name.
